src/uct/ib/ud/base/ud_ep.c

```diff
@@ -1063,6 +1068,69 @@ ucs_status_t uct_ud_ep_flush(uct_ep_h ep_h, unsigned flags,
     return status;
 }
 
+static uct_ud_send_skb_t *uct_ud_ep_prepare_check(uct_ud_ep_t *ep)
+{
+    uct_ud_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_ud_iface_t);
+    uct_ud_send_skb_t *skb;
+    uct_ud_neth_t *neth;
+
+    ucs_assert_always(ep->dest_ep_id != UCT_UD_EP_NULL_ID);
+    ucs_assert_always(ep->ep_id != UCT_UD_EP_NULL_ID);
+
+    skb = uct_ud_iface_get_tx_skb(iface, ep);
+    if (skb == NULL) {
+        return NULL;
+    }
+
+    neth              = skb->neth;
+    neth->packet_type = ep->dest_ep_id | UCT_UD_PACKET_FLAG_CHECK;
+    uct_ud_neth_init_data(ep, neth);
+    uct_ud_neth_ack_req(ep, neth);
+
+    skb->len = sizeof(*neth);
+    uct_ud_ep_ctl_op_del(ep, UCT_UD_EP_OP_CHECK);
+    return skb;
+}
+
+ucs_status_t uct_ud_ep_check(uct_ep_h tl_ep, unsigned flags, uct_completion_t *comp)
+{
+    uct_ud_ep_t *ep       = ucs_derived_of(tl_ep, uct_ud_ep_t);
+    uct_ud_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_ud_iface_t);
+    uct_ud_send_skb_t *skb;
+
+    UCT_EP_KEEPALIVE_CHECK_PARAM(flags, comp);
+
+    uct_ud_enter(iface);
+
+    if (uct_ud_ep_ctl_op_check(ep, UCT_UD_EP_OP_CHECK)) {
+        /* keepalive request is in pending queue and will be
+         * processed when resources are available */
+        goto out;
+    }
+
+    if (!uct_ud_ep_is_connected(ep) ||
+        !uct_ud_ep_is_last_ack_received(ep)) {
+        /* in case if no TX resources are available then there is at least one
+         * signaled operation which provides actual peer status, in this case just
+         * return without any actions */
+        goto out;
+    }
+
+    skb = uct_ud_ep_prepare_check(ep);
+    if (skb == NULL) {
+        /* schedule keepalive request to pending queue for futher sending */
+        uct_ud_ep_ctl_op_add(iface, ep, UCT_UD_EP_OP_CHECK);
+        goto out;
+    }
+
+    uct_ud_iface_send_ctl(iface, ep, skb, NULL, 0, 0, 1);
+    uct_ud_iface_complete_tx_skb(iface, ep, skb);
```

Comment:
just need to add ctl_op, and let it be sent from progress

Reply:
done